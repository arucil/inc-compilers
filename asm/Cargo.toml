[package]
name = "asm"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "1.8.1"
enum-iterator = "0.8.1"
num-derive = "0.3.3"
num-traits = "0.2.15"
support = { path = "../support" }
