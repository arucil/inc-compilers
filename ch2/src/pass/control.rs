use ast::{Program, Exp, IdxVar};
use std::fmt::{self, Write};
use indexmap::IndexSet;
use support::{WritePretty, Range};

pub struct CProgram {
  pub info: CInfo,
  pub body: Vec<(String, CTail)>,
}

pub struct CInfo {
  pub locals: IndexSet<IdxVar>,
}

pub enum CTail {
  Return(CExp),
  Seq(CStmt, Box<CTail>)
}

pub enum CStmt {
  Assign {
    var: IdxVar,
    exp: CExp,
  }
}

pub enum CExp {
  Atom(CAtom),
  Prim(CPrim),
}

pub enum CPrim {
  Read,
  Neg(CAtom),
  Add(CAtom, CAtom),
}

pub enum CAtom {
  Int(i64),
  Var(IdxVar)
}

impl CProgram {
  pub fn to_string_pretty(&self) -> String {
    let mut buf = String::new();
    self.info.write(&mut buf).unwrap();
    for (label, tail) in &self.body {
      writeln!(&mut buf, "{}:", label).unwrap();
      tail.write(&mut buf).unwrap();
    }
    buf
  }
}

impl WritePretty for CInfo {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

impl WritePretty for CTail {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Return(exp) => {
        write!(f, "    return ")?;
        exp.write(f)
      }
      Self::Seq(stmt, tail) => {
        write!(f, "    ")?;
        stmt.write(f)?;
        write!(f, "\n")?;
        tail.write(f)
      }
    }
  }
}

impl WritePretty for CStmt {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Assign { var, exp } => {
        write!(f, "{:?} = ", var)?;
        exp.write(f)
      }
    }
  }
}

impl WritePretty for CExp {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Atom(atom) => atom.write(f),
      Self::Prim(prim) => prim.write(f),
    }
  }
}

impl WritePretty for CPrim {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Add(arg1, arg2) => {
        write!(f, "(+ ")?;
        arg1.write(f)?;
        write!(f, " ")?;
        arg2.write(f)?;
        write!(f, ")")
      }
      Self::Neg(arg) => {
        write!(f, "(- ")?;
        arg.write(f)?;
        write!(f, ")")
      }
      Self::Read => {
        write!(f, "(read)")
      }
    }
  }
}

impl WritePretty for CAtom {
  fn write(&self, f: &mut impl Write) -> fmt::Result {
    match self {
      Self::Int(n) => write!(f, "{}", n),
      Self::Var(n) => write!(f, "{:?}", n),
    }
  }
}

pub fn explicate_control(mut prog: Program<IdxVar>) -> CProgram {
  let locals = collect_locals(&prog);
  CProgram {
    info: CInfo {
      locals,
    },
    body: vec![
      ("start".to_owned(), explicate_tail(prog.body.pop().unwrap().1))
    ]
  }
}

fn collect_locals(
  prog: &Program<IdxVar>,
) -> IndexSet<IdxVar> {
  let mut locals = IndexSet::new();
  for (_, exp) in &prog.body {
    collect_exp_locals(exp, &mut locals);
  }
  locals
}

fn collect_exp_locals(
  exp: &Exp<IdxVar>,
  locals: &mut IndexSet<IdxVar>,
) {
  match exp {
    Exp::Int(_) => {}
    Exp::Var(var) => {
      locals.insert(var.clone());
    }
    Exp::Str(_) => {}
    Exp::Let { var, init, body } => {
      locals.insert(var.1.clone());
      collect_exp_locals(&init.1, locals);
      collect_exp_locals(&body.1, locals);
    }
    Exp::Prim { op: _, args } => {
      for (_, arg) in args {
        collect_exp_locals(arg, locals);
      }
    }
    _ => unimplemented!()
  }
}

fn explicate_tail(exp: Exp<IdxVar>) -> CTail {
  match exp {
    exp@(Exp::Int(_) | Exp::Var(_)) => {
      CTail::Return(CExp::Atom(atom(exp)))
    }
    ast::Exp::Prim { op: (_, op), args } => {
      CTail::Return(CExp::Prim(prim(op, args)))
    }
    ast::Exp::Let { var, init, body } => {
      explicate_assign(var.1, init.1, explicate_tail(body.1))
    }
    exp => unimplemented!("unsupported form {:?}", exp)
  }
}

fn explicate_assign(
  var: IdxVar,
  init: ast::Exp<IdxVar>,
  cont: CTail,
) -> CTail {
  match init {
    exp@(Exp::Int(_) | Exp::Var(_)) => {
      let assign = CStmt::Assign {
        var,
        exp: CExp::Atom(atom(exp)),
      };
      CTail::Seq(assign, box cont)
    }
    Exp::Prim { op: (_, op), args } => {
      let prim = prim(op, args);
      let assign = CStmt::Assign {
        var,
        exp: CExp::Prim(prim),
      };
      CTail::Seq(assign, box cont)
    }
    Exp::Let { var: (_, var1), init: box (_, init1), body } => {
      explicate_assign(var1, init1, explicate_assign(var, body.1, cont))
    }
    exp => unimplemented!("unsupported form {:?}", exp)
  }
}

fn atom(exp: Exp<IdxVar>) -> CAtom {
  match exp {
    Exp::Int(n) => CAtom::Int(n),
    Exp::Var(var) => CAtom::Var(var),
    exp => unreachable!("{:?}", exp),
  }
}

fn prim(op: &str, mut args: Vec<(Range, Exp<IdxVar>)>) -> CPrim {
  match op {
    "read" => CPrim::Read,
    "-" => CPrim::Neg(atom(args.pop().unwrap().1)),
    "+" => CPrim::Add(
      atom(args[0].1.clone()),
      atom(args.pop().unwrap().1),
    ),
    _ => unreachable!("{}", op)
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn let_in_init() {
    let prog = parse(r#"(let ([y (let ([x 20]) (+ x (let ([x 22]) x)))]) y)"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_prims() {
    let prog = parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#).unwrap();
    let prog = super::super::uniquify::uniquify(prog).unwrap();
    let prog = super::super::anf::anf(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}