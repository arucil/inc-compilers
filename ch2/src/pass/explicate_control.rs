use asm::Label;
use ast::{Exp, ExpKind, IdxVar, Program};
use control::{CAtom, CExp, CPrim, CProgram, CStmt, CTail};
use indexmap::IndexSet;
use std::fmt::{self, Debug, Formatter};

pub struct CInfo {
  pub locals: IndexSet<IdxVar>,
}

impl Debug for CInfo {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

pub fn explicate_control(mut prog: Program<IdxVar>) -> CProgram<CInfo> {
  let locals = collect_locals(&prog);
  CProgram {
    info: CInfo { locals },
    body: vec![(Label::Start, explicate_tail(prog.body.pop().unwrap()))],
  }
}

fn explicate_tail(exp: Exp<IdxVar>) -> CTail {
  match exp.kind {
    ExpKind::Int(_) | ExpKind::Var(_) => CTail::Return(CExp::Atom(atom(exp))),
    ExpKind::Prim { op: (_, op), args } => {
      CTail::Return(CExp::Prim(prim(op, args)))
    }
    ExpKind::Let { var, init, body } => {
      explicate_assign(var.1, *init, explicate_tail(*body))
    }
    exp => unimplemented!("unsupported form {:?}", exp),
  }
}

fn explicate_assign(var: IdxVar, init: ast::Exp<IdxVar>, cont: CTail) -> CTail {
  match init.kind {
    ExpKind::Int(_) | ExpKind::Var(_) => {
      let assign = CStmt::Assign {
        var,
        exp: CExp::Atom(atom(init)),
      };
      CTail::Seq(assign, box cont)
    }
    ExpKind::Prim { op: (_, op), args } => {
      let prim = prim(op, args);
      let assign = CStmt::Assign {
        var,
        exp: CExp::Prim(prim),
      };
      CTail::Seq(assign, box cont)
    }
    ExpKind::Let {
      var: (_, var1),
      init: box init1,
      body,
    } => explicate_assign(var1, init1, explicate_assign(var, *body, cont)),
    exp => unimplemented!("unsupported form {:?}", exp),
  }
}

fn atom(exp: Exp<IdxVar>) -> CAtom {
  match exp.kind {
    ExpKind::Int(n) => CAtom::Int(n),
    ExpKind::Var(var) => CAtom::Var(var),
    exp => unreachable!("{:?}", exp),
  }
}

fn prim(op: &str, mut args: Vec<Exp<IdxVar>>) -> CPrim {
  match op {
    "read" => CPrim::Read,
    "-" => {
      if args.len() == 1 {
        CPrim::Neg(atom(args.pop().unwrap()))
      } else {
        CPrim::Sub(atom(args[0].clone()), atom(args.pop().unwrap()))
      }
    }
    "+" => CPrim::Add(atom(args[0].clone()), atom(args.pop().unwrap())),
    _ => unreachable!("{}", op),
  }
}

fn collect_locals<TYPE>(prog: &Program<IdxVar, TYPE>) -> IndexSet<IdxVar> {
  let mut locals = IndexSet::new();
  for exp in &prog.body {
    collect_exp_locals(exp, &mut locals);
  }
  locals
}

fn collect_exp_locals<TYPE>(
  exp: &Exp<IdxVar, TYPE>,
  locals: &mut IndexSet<IdxVar>,
) {
  match &exp.kind {
    ExpKind::Int(_)
    | ExpKind::Var(_)
    | ExpKind::Str(_)
    | ExpKind::Bool(_)
    | ExpKind::Void
    | ExpKind::Get(_) => {}
    ExpKind::Let { var, init, body } => {
      locals.insert(var.1.clone());
      collect_exp_locals(&*init, locals);
      collect_exp_locals(&*body, locals);
    }
    ExpKind::Prim { op: _, args } => {
      for arg in args {
        collect_exp_locals(arg, locals);
      }
    }
    ExpKind::If { cond, conseq, alt } => {
      collect_exp_locals(&*cond, locals);
      collect_exp_locals(&*conseq, locals);
      collect_exp_locals(&*alt, locals);
    }
    ExpKind::Set { var: _, exp } => {
      collect_exp_locals(&*exp, locals);
    }
    ExpKind::Begin { seq, last } => {
      for exp in seq {
        collect_exp_locals(&*exp, locals);
      }
      collect_exp_locals(&*last, locals);
    }
    ExpKind::While { cond, body } => {
      collect_exp_locals(&*cond, locals);
      collect_exp_locals(&*body, locals);
    }
    ExpKind::Print(args) => {
      for arg in args {
        collect_exp_locals(arg, locals);
      }
    }
    ExpKind::NewLine => {}
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use insta::assert_snapshot;

  #[test]
  fn let_in_init() {
    let prog = ast::parse(
      r#"
        (let
          ([y (let ([x 20])
                (+ x
                   (let ([x 22]) x)))])
          y)
        "#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn nested_prims() {
    let prog = ast::parse(
      r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#,
    )
    .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
