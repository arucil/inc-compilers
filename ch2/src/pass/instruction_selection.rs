use super::explicate_control::CInfo;
use asm::{Arg, Block, ByteReg, Instr, Label, Program, Reg};
use ast::{IdxVar, Type};
use control::*;
use indexmap::{IndexMap, IndexSet};
use std::collections::BTreeSet;
use std::fmt::{self, Debug, Formatter};

pub struct Info {
  pub locals: IndexSet<IdxVar>,
}

impl Debug for Info {
  fn fmt(&self, f: &mut Formatter) -> fmt::Result {
    writeln!(f, "locals: {:?}\n", self.locals)
  }
}

impl From<CInfo> for Info {
  fn from(info: CInfo) -> Self {
    Self {
      locals: info.locals,
    }
  }
}

pub fn select_instruction<OldInfo, NewInfo>(
  prog: CProgram<OldInfo>,
  use_heap: bool,
) -> Program<NewInfo, IdxVar>
where
  NewInfo: From<OldInfo>,
{
  let mut codegen = CodeGen::new(use_heap);
  let blocks = prog
    .body
    .into_iter()
    .map(|(label, tail)| (label, codegen.tail_block(tail)))
    .collect();
  let result = codegen.finish();
  Program {
    info: prog.info.into(),
    blocks,
    constants: result.constants,
    externs: result.externs,
  }
}

#[derive(Default)]
pub struct CodeGen {
  code: Vec<Instr<IdxVar>>,
  constants: IndexMap<String, String>,
  externs: BTreeSet<String>,
  use_heap: bool,
}

pub struct CodeGenResult {
  pub constants: IndexMap<String, String>,
  pub externs: BTreeSet<String>,
}

impl CodeGen {
  pub fn new(use_heap: bool) -> Self {
    Self {
      code: vec![],
      constants: IndexMap::new(),
      externs: BTreeSet::new(),
      use_heap,
    }
  }

  pub fn finish(self) -> CodeGenResult {
    CodeGenResult {
      constants: self.constants,
      externs: self.externs,
    }
  }

  pub fn tail_block(&mut self, tail: CTail) -> Block<IdxVar> {
    self.code.clear();
    self.tail_instructions(tail);
    Block {
      global: false,
      code: std::mem::take(&mut self.code),
    }
  }

  fn tail_instructions(&mut self, mut tail: CTail) {
    loop {
      match tail {
        CTail::Return(exp) => {
          self.exp_instructions(Arg::Reg(Reg::Rax), exp);
          self.code.push(Instr::Jmp(Label::Conclusion));
          return;
        }
        CTail::Seq(stmt, new_tail) => {
          self.stmt_instructions(stmt);
          tail = *new_tail;
        }
        // ch4
        CTail::Goto(label) => {
          self.code.push(Instr::Jmp(label));
          return;
        }
        CTail::If {
          cmp,
          lhs,
          rhs,
          conseq,
          alt,
        } => {
          let lhs = atom_to_arg(lhs);
          let rhs = atom_to_arg(rhs);
          self.code.push(Instr::Cmp {
            dest: lhs,
            src: rhs,
          });
          self.code.push(Instr::JumpIf {
            cmp: cmp.into(),
            label: conseq,
          });
          self.code.push(Instr::Jmp(alt));
          return;
        }
      }
    }
  }

  fn stmt_instructions(&mut self, stmt: CStmt) {
    match stmt {
      CStmt::Assign { var, exp } => self.exp_instructions(Arg::Var(var), exp),
      // ch4
      CStmt::Print { val, ty: Type::Int } => {
        self.atom_instructions(Arg::Reg(Reg::Rdi), val);
        let label = "rt_print_int".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 0 });
      }
      CStmt::Print {
        val,
        ty: Type::Bool,
      } => {
        self.atom_instructions(Arg::Reg(Reg::Rdi), val);
        let label = "rt_print_bool".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 0 });
      }
      CStmt::Print { val, ty: Type::Str } => {
        self.atom_instructions(Arg::Reg(Reg::Rdi), val);
        let label = "rt_print_str".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 0 });
      }
      CStmt::Print { .. } => unreachable!(),
      CStmt::NewLine => {
        let label = "rt_print_newline".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 0 })
      }
      CStmt::Read => {
        let label = "rt_read_int".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 0 })
      }
      CStmt::VecSet {
        vec,
        fields_before,
        val,
      } => {
        let vec = atom_to_arg(vec);
        // TODO remove R11 requirement
        self.code.push(Instr::Mov {
          src: vec,
          dest: Arg::Reg(Reg::R11),
        });
        self.atom_instructions(
          Arg::Deref(Reg::R11, calc_field_offset(&fields_before)),
          val,
        );
      }
    }
  }

  fn exp_instructions(&mut self, target: Arg<IdxVar>, exp: CExp) {
    match exp {
      CExp::Atom(atom) => self.atom_instructions(target, atom),
      CExp::Prim(prim) => self.prim_instructions(target, prim),
    }
  }

  fn prim_instructions(&mut self, target: Arg<IdxVar>, prim: CPrim) {
    match prim {
      CPrim::Read => {
        let label = "rt_read_int".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 0 });
        self.code.push(Instr::Mov {
          src: Arg::Reg(Reg::Rax),
          dest: target,
        });
      }
      CPrim::Neg(atom) => {
        self.atom_instructions(target.clone(), atom);
        self.code.push(Instr::Neg(target));
      }
      CPrim::Add(atom1, atom2) => {
        self.atom_instructions(target.clone(), atom1);
        let arg = atom_to_arg(atom2);
        self.code.push(Instr::Add {
          src: arg,
          dest: target,
        });
      }
      CPrim::Sub(atom1, atom2) => {
        self.atom_instructions(target.clone(), atom1);
        let arg = atom_to_arg(atom2);
        self.code.push(Instr::Sub {
          src: arg,
          dest: target,
        });
      }
      // ch4
      CPrim::Not(atom) => {
        self.atom_instructions(target.clone(), atom);
        self.code.push(Instr::Xor {
          src: Arg::Imm(1),
          dest: target,
        });
      }
      CPrim::Cmp(cmp, atom1, atom2) => {
        let arg1 = atom_to_arg(atom1);
        let arg2 = atom_to_arg(atom2);
        self.code.push(Instr::Cmp {
          dest: arg1,
          src: arg2,
        });
        self.code.push(Instr::SetIf {
          cmp: cmp.into(),
          dest: Arg::ByteReg(ByteReg::Al),
        });
        self.code.push(Instr::Movzx {
          src: Arg::ByteReg(ByteReg::Al),
          dest: target,
        });
      }
      // ch6
      CPrim::Vector(fields) => {
        let mut size = 8;
        let mut non_unit_fields = fields.len() as i64;
        let mut ptr_mask = 0i64;
        let mut field_offsets = vec![];
        for (i, (_, ty)) in fields.iter().enumerate() {
          let offset = size as i32;
          match ty {
            Type::Void => {
              non_unit_fields -= 1;
            }
            Type::Int => size += 8,
            Type::Bool => size += 8,
            Type::Str => {
              size += 8;
              ptr_mask |= 1 << i;
            }
            Type::Vector(_) => {
              size += 8;
              ptr_mask |= 1 << i;
            }
            Type::Alias(_) => todo!(),
          }
          field_offsets.push(offset);
        }
        ptr_mask >>= 1;
        self.code.push(Instr::Mov {
          src: Arg::Imm(size),
          dest: Arg::Reg(Reg::Rdi),
        });
        self.code.push(Instr::Mov {
          src: Arg::Reg(Reg::R15),
          dest: Arg::Reg(Reg::Rsi),
        });
        let label = "rt_allocate".to_owned();
        self.externs.insert(label.clone());
        self.code.push(Instr::Call { label, arity: 2 });
        self.code.push(Instr::Mov {
          src: Arg::Reg(Reg::Rax),
          dest: Arg::Reg(Reg::R11),
        });
        self.code.push(Instr::Mov {
          src: Arg::Imm(ptr_mask << 9 | non_unit_fields << 3 | 1),
          dest: Arg::Deref(Reg::R11, 0),
        });
        for ((field, ty), offset) in fields.into_iter().zip(field_offsets) {
          if ty == Type::Void {
            continue;
          }
          self.atom_instructions(Arg::Deref(Reg::R11, offset), field);
        }
        self.code.push(Instr::Mov {
          src: Arg::Reg(Reg::R11),
          dest: target,
        })
      }
      CPrim::VecRef { vec, fields_before } => {
        self.atom_instructions(Arg::Reg(Reg::R11), vec);
        self.code.push(Instr::Mov {
          src: Arg::Deref(Reg::R11, calc_field_offset(&fields_before)),
          dest: target,
        });
      }
      CPrim::VecLen(arg) => {
        self.atom_instructions(Arg::Reg(Reg::R11), arg);
        self.code.push(Instr::Mov {
          src: Arg::Deref(Reg::R11, 0),
          dest: target.clone(),
        });
        self.code.push(Instr::Shr {
          src: target.clone(),
          count: Arg::Imm(3),
        });
        self.code.push(Instr::And {
          src: Arg::Imm(0b111111),
          dest: target,
        });
      }
    }
  }

  fn atom_instructions(&mut self, target: Arg<IdxVar>, atom: CAtom) {
    match atom {
      CAtom::Int(_) | CAtom::Var(_) | CAtom::Bool(_) => {
        self.code.push(Instr::Mov {
          src: atom_to_arg(atom),
          dest: target,
        });
      }
      // ch5
      CAtom::Void => {}
      CAtom::Str(s) => {
        let label = format!("const_{}", self.constants.len() + 1);
        let len = s.len();
        self.constants.insert(label.clone(), s);
        if self.use_heap {
          self.code.push(Instr::Mov {
            src: Arg::Imm(len as i64),
            dest: Arg::Reg(Reg::Rdi),
          });
          self.code.push(Instr::Mov {
            src: Arg::Label(label),
            dest: Arg::Reg(Reg::Rsi),
          });
          self.code.push(Instr::Mov {
            src: Arg::Reg(Reg::R15),
            dest: Arg::Reg(Reg::Rdx),
          });
          let label = "rt_new_string".to_owned();
          self.externs.insert(label.clone());
          self.code.push(Instr::Call { label, arity: 3 });
          self.code.push(Instr::Mov {
            src: Arg::Reg(Reg::Rax),
            dest: target,
          });
        } else {
          self.code.push(Instr::Mov {
            src: Arg::Label(label),
            dest: target,
          });
        }
      }
    }
  }
}

fn calc_field_offset(fields_before: &[Type]) -> i32 {
  let mut offset = 8;
  for field in fields_before {
    offset += match field {
      Type::Void => 0,
      Type::Int => 8,
      Type::Bool => 8,
      Type::Str => 8,
      Type::Vector(_) => 8,
      Type::Alias(_) => todo!(),
    };
  }
  offset
}

fn atom_to_arg(atom: CAtom) -> Arg<IdxVar> {
  match atom {
    CAtom::Int(n) => Arg::Imm(n),
    CAtom::Var(var) => Arg::Var(var),
    CAtom::Bool(true) => Arg::Imm(1),
    CAtom::Bool(false) => Arg::Imm(0),
    CAtom::Void => unreachable!(),
    CAtom::Str(_) => unreachable!(),
  }
}

#[cfg(test)]
mod tests {
  use super::super::*;
  use super::*;
  use ast::*;
  use insta::assert_snapshot;

  #[test]
  fn nested_prims() {
    let prog =
      parse(r#"(let ([x (read)] [y (+ 2 3)]) (+ (- (read)) (+ y (- 2))))"#)
        .unwrap();
    let prog = uniquify::uniquify(prog).unwrap();
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = select_instruction::<_, Info>(prog, false);
    assert_snapshot!(result.to_string_pretty());
  }
}
