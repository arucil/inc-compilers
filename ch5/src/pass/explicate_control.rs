#[cfg(test)]
mod tests {
  use super::super::*;
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use insta::assert_snapshot;

  #[test]
  fn begin() {
    let prog = ast::parse(
      r#"
(let ([x 1] [y (read)])
  (while (> x 10)
    (set! x (begin (print "x=" x) (+ x (- y 1))))
    (print (> x (- y 1))))
  (begin 1 2 3)
  (let ([k (+ x 1)])
    (print k)
    k)
  (+ 7 (+ (read) (if (begin (set! y 37) (not (eq? y x))) 11 23)))
  (+ x y)
  (void))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn assign_void() {
    let prog = ast::parse(
      r#"(let ([x (void)] [y 1]) (set! x (print)) (set! x (set! y 10)))"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
