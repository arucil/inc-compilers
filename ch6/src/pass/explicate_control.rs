#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use insta::assert_snapshot;

  #[test]
  fn vector() {
    let prog = ast::parse(
      r#"
(let ([x (void)] [y (vector 1 #f (void) (vector (set! x x)) "abc")])
  (set! x (vector-set! y 1 (not (vector-ref y 1))))
  (set! x (vector-ref y 2))
  (vector-set! y 0 (vector-length y))
  (vector-set! y 2 (vector-ref (vector-ref y 3) 0)))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn r#struct() {
    let prog = ast::parse(
      r#"
(define-struct my-void (x Void))
(define-struct my-comp (x Int) (y Bool) (z Void) (u my-void) (v Str))
(let ([x (void)] [y (my-comp 1 #f (void) (my-void (set! x x)) "abc")])
  (set! x (set-my-comp-y! y (not (my-comp-y y))))
  (set! x (my-comp-z y))
  (set-my-comp-x! y (vector-length y))
  (set-my-comp-z! y (my-void-x (my-comp-u y))))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
