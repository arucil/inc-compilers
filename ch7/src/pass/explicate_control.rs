#[cfg(test)]
mod tests {
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use ch5::pass::typecheck;
  use ch6::pass::array_bounds;
  use insta::assert_snapshot;

  #[test]
  fn call_in_cond() {
    let prog = ast::parse(
      r#"
(define (foo [x (Int Bool Void)]) : Bool
  (if (vector-ref x 1)
    (not (vector-ref x 1))
    (> 3 2)))
(print
  (if (foo (vector (read) #t (void)))
    (begin
      (print 3)
      #t)
    #f))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn call_effect() {
    let prog = ast::parse(
      r#"
(define (foo [x (Int Bool Void)]) : Bool
  (if (vector-ref x 1)
    (not (vector-ref x 1))
    (> 3 2)))
(foo (vector 1 (foo (vector 1 #f (void))) (newline)))
(void)
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = array_bounds::insert_bounds_check(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn void_in_param() {
    let prog = ast::parse(
      r#"
(define (foo [x Int] [y Void] [z Str]) : Bool
  (set! y (void))
  (let ([t (void)])
    (set! t y))
  y
  (print z)
  (not (eq? x 3)))
(void)
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = array_bounds::insert_bounds_check(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn tailcall() {
    let prog = ast::parse(
      r#"
(define (foo [x Int] [y Void] [z Str]) : Bool
  (set! y (void))
  (let ([t (void)])
    (set! t y))
  y
  (print z)
  (foo (+ x 1) y "z"))
(void)
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = array_bounds::insert_bounds_check(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
