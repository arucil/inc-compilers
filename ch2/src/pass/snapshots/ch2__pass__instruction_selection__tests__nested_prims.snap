---
source: ch2/src/pass/instruction_selection.rs
expression: result.to_string_pretty()
---
locals: {x.0, y.1, (tmp).0, (tmp).1, (tmp).2, (tmp).3}

constants: {}

.start:
    call rt_read_int
    mov x.0, rax
    mov y.1, 2
    add y.1, 3
    call rt_read_int
    mov (tmp).0, rax
    mov (tmp).1, (tmp).0
    neg (tmp).1
    mov (tmp).2, 2
    neg (tmp).2
    mov (tmp).3, y.1
    add (tmp).3, (tmp).2
    mov rax, (tmp).1
    add rax, (tmp).3
    jmp .epilogue

